Sources/PackageGraph/DependencyResolver.swift

```diff
@@ -1094,3 +1153,139 @@ public class DependencyResolver<
         return container
     }
 }
+
+/// The resolver debugger.
+///
+/// Finds the constraints which results in graph being unresolvable.
+private struct ResolverDebugger<
+    Provider: PackageContainerProvider,
+    Delegate: DependencyResolverDelegate
+> where Provider.Container.Identifier == Delegate.Identifier {
+
+    typealias Identifier = Provider.Container.Identifier
+    typealias Constraint = PackageContainerConstraint<Identifier>
+
+    /// Reference to the resolver.
+    unowned let resolver: DependencyResolver<Provider, Delegate>
+
+    /// Create a new debugger.
+    init(_ resolver: DependencyResolver<Provider, Delegate>) {
+        self.resolver = resolver
+    }
+
+    // Returns the constraints which should be removed in order to make the
+    // graph resolvable.
+    func debug(
+        dependencies: [Constraint],
+        pins: [Constraint]
+    ) -> (dependencies: [Constraint], pins: [Constraint]) {
+        // Put the resolver in incomplete mode to avoid cloning new repositories.
+        resolver.isInIncompleteMode = true
+
+        let deltaAlgo = DeltaAlgorithm<ResolverChange>()
+        let allPackages = Set(dependencies.map({ $0.identifier }))
+
+        // Compute the set of changes.
+        let changes = dependencies.map({ ResolverChange.allowPackage($0.identifier) })
+                    + pins.map({ ResolverChange.allowPin($0.identifier) })
+
+        // Run the delta debugging algorithm.
+        let badChanges = deltaAlgo.run(changes: Set(changes)){ changes in
```

Comment:
This algorithm can be exponential in practice, so I propose that we should have a maximum time limit we are willing to try to debug (10s seems like a reasonable first choice), and after that we just throw an error saying we couldn't diagnose the problem.